### PR TITLE
feat: Add support for query API to influxdb_client

### PR DESCRIPTION
**Pull request by aakashhemadri**
POST /query/ast
GET   /query/suggestions
GET   /query/suggestions/{name}
POST /query/analyze
POST /query

Implemented all query APIs, needed models, tests, and examples

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: influxdb2_client/src/models/ast/dialect.rs [hunk 1]
```diff
@@ -0,0 +1,52 @@
+//! Dialect
+
+use serde::{Deserialize, Serialize};
+
+/// Dialect are options to change the default CSV output format; https://www.w3.org/TR/2015/REC-tabular-metadata-20151217/#dialect-descriptions
+#[derive(Clone, Debug, PartialEq, Default, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Dialect {
```
Comment (carols10cents) [inline influxdb2_client/src/models/ast/dialect.rs:8]: Hm, there are some fields in the spec that aren't here-- are they not supported by InfluxDB (I'm not sure)? I don't see `doubleQuote`, `encoding`, `headerRowCount`, `lineTerminators`, among others...?

**Review by carols10cents — CHANGES_REQUESTED** [review]
I found a few things I'd like to see fixed; there might be more but I keep getting pulled away to other things and I don't want to forget to submit this :) So I might have more comments next round....

File: influxdb2_client/src/api/query.rs [hunk 2]
```diff
@@ -0,0 +1,344 @@
+//! Onboarding/Setup
+//!
+//! Initate and start onboarding process of InfluxDB server.
```
[the remark this answers is not in the document]
Comment (aakashhemadri) [reply influxdb2_client/src/api/query.rs:3]: Sure! 😅

File: influxdb2_client/src/api/query.rs [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+                Http { status, text }.fail()?
+            }
+        }
+    }
+
+    /// Query
+    pub async fn query(
+        &self,
+        org: &str,
+        org_id: &str,
```
[the remark this answers is not in the document]
Comment (aakashhemadri) [reply influxdb2_client/src/api/query.rs:65]: You are right as to both being optional, I have also looked into the C# Client, Seems to be the same as Go, just org being used. I will make this change once you give me confirmation

File: influxdb2_client/src/models/ast/call_expression.rs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+pub struct CallExpression {
+    /// Type of AST node
+    #[serde(rename = "type", skip_serializing_if = "Option::is_none")]
+    pub _type: Option<String>,
+    /// Callee
+    #[serde(rename = "callee", skip_serializing_if = "Option::is_none")]
+    pub callee: Option<Box<crate::models::ast::Expression>>,
+    /// Function arguments
+    #[serde(rename = "arguments", skip_serializing_if = "Option::is_none")]
+    pub arguments: Option<Vec<crate::models::ast::Expression>>,
```
[the remark this answers is not in the document]
Comment (aakashhemadri) [reply influxdb2_client/src/models/ast/call_expression.rs:16]: Ah I missed this file altogether 😅

File: influxdb2_client/src/api/query.rs [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+    ) -> Result<String, RequestError> {
+        let req_url = format!("{}/api/v2/query", self.url);
+
+        let response = self
+            .request(Method::POST, &req_url)
+            .header("Accepting-Encoding", "identity")
+            .header("Content-Type", "application/json")
+            .query(&[("org", &org), ("orgID", &org_id)])
+            .body(
+                serde_json::to_string(&query.unwrap_or_else(|| Query::new("".to_string())))
```
[the remark this answers is not in the document]
Comment (aakashhemadri) [reply influxdb2_client/src/api/query.rs:76]: Ah your fix is cleaner :)

File: influxdb2_client/src/models/ast/dialect.rs (hunk 1 above)
Comment (aakashhemadri) [reply influxdb2_client/src/models/ast/dialect.rs:8]: Ah not sure about this, This is what the swagger spec generated, I assumed it was correct. Can you point me to the specification itself? I did find [this..](https://docs.influxdata.com/influxdb/v2.0/reference/flux/language/) but can't seem to find anything else.

File: influxdb2_client/src/api/query.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (carols10cents) [reply influxdb2_client/src/api/query.rs:65]: Yes, please!

File: influxdb2_client/src/models/ast/dialect.rs (hunk 1 above)
Comment (carols10cents) [reply influxdb2_client/src/models/ast/dialect.rs:8]: Oh the spec is at the link in the doc comment: https://www.w3.org/TR/2015/REC-tabular-metadata-20151217/#dialect-descriptions

If this is what the Influx swagger generates though, that's fine then.

File: influxdb2_client/src/models/ast/dict_item.rs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+use serde::{Deserialize, Serialize};
+
+/// A key/value pair in a dictionary
+#[derive(Clone, Debug, PartialEq, Default, Serialize, Deserialize)]
+pub struct DictItem {
+    /// Type of AST node
+    #[serde(rename = "type", skip_serializing_if = "Option::is_none")]
+    pub r#type: Option<String>,
+    /// Key
+    #[serde(rename = "key", skip_serializing_if = "Option::is_none")]
```
Comment (carols10cents) [inline influxdb2_client/src/models/ast/dict_item.rs:12]: This rename isn't needed

File: influxdb2_client/src/models/ast/dict_item.rs [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    pub key: Option<crate::models::ast::Expression>,
+    /// Value
+    #[serde(rename = "val", skip_serializing_if = "Option::is_none")]
```
Comment (carols10cents) [inline influxdb2_client/src/models/ast/dict_item.rs:15]: This rename isn't needed

File: influxdb2_client/src/models/ast/expression.rs [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    pub right: Option<Box<crate::models::ast::Expression>>,
+    /// Parent Expression
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub callee: Option<Box<crate::models::ast::Expression>>,
+    /// Function arguments
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub arguments: Vec<crate::models::ast::Expression>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    /// Test Expr
+    pub test: Option<Box<crate::models::ast::Expression>>,
```
Comment (carols10cents) [inline influxdb2_client/src/models/ast/expression.rs:37]: I got used to reading:

```
/// Doc comment
#[annotations]
pub field: ...
```

but this one is:

```
#[annotations]
/// Doc comment
pub field: ...
```

Is there any reason for the difference or could this be made consistent with the other fields?

File: influxdb2_client/src/models/ast/expression.rs [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    pub properties: Vec<crate::models::ast::Property>,
+    /// Expression
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub expression: Option<Box<crate::models::ast::Expression>>,
+    /// Argument
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub argument: Option<Box<crate::models::ast::Expression>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    /// Call Expr
+    pub call: Option<crate::models::ast::CallExpression>,
```
Comment (carols10cents) [inline influxdb2_client/src/models/ast/expression.rs:67]: Same ordering inconsistency here

File: influxdb2_client/src/models/query.rs [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    pub r#extern: Option<File>,
+    /// Query script to execute.
+    pub query: String,
+    /// The type of query. Must be \"flux\".
+    #[serde(rename = "type", skip_serializing_if = "Option::is_none")]
+    pub r#type: Option<Type>,
+    /// Dialect
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub dialect: Option<crate::models::ast::Dialect>,
+    /// Specifies the time that should be reported as \"now\" in the query.
```
Comment (carols10cents) [inline influxdb2_client/src/models/query.rs:22]: Do these double quotes really need to be escaped in this doc comment? I'm guessing that was something the swagger code generator did... could you remove the escaping please?

**Review by carols10cents — CHANGES_REQUESTED** [review]
This is getting close! Just a few more tiny little picky changes. Thanks!

File: influxdb2_client/src/models/ast/expression.rs (hunk 8 above)
Comment (aakashhemadri) [reply influxdb2_client/src/models/ast/expression.rs:37]: Ah no just an error on my end

**Review by carols10cents — APPROVED** [review]
This looks great now, thank you!!

**Comment by alamb** [comment]
Thanks @aakashhemadri !

**Pull request merged**